app/src/main/java/io/github/cmput301f19t19/legendary_fiesta/FirebaseHelper.java

```diff
@@ -60,6 +60,123 @@ public void checkUserExists(String name, final FirebaseCallback<QuerySnapshot> c
                 .addOnFailureListener(callback::onFailure);
     }
 
+    /**
+     * check if there is already a friend request concerning the two users
+     *
+     * @param fromUID
+     * @param toUID
+     * @param callback
+     */
+    public void checkRequestExists(String fromUID, String toUID, final FirebaseCallback<Boolean> callback) {
+        if (fromUID.equals(toUID)) {
+            callback.onFailure(new Exception("Cannot send yourself friend requests"));
+            return;
+        }
+        db.collection("requests").whereEqualTo("from",fromUID).whereEqualTo("to", toUID)
+            .get()
+                .addOnSuccessListener(queryDocumentSnapshots -> callback.onSuccess(!queryDocumentSnapshots.isEmpty()))
+                .addOnFailureListener(callback::onFailure);
+    }
+
+    /**
+     * @param UID
+     * @param callback
+     */
+    public void getPendingRequests(String UID, final FirebaseCallback<List<FriendRequest>> callback) {
+        db.collection("requests").whereEqualTo("to", UID).whereEqualTo("status", false)
+                .get().addOnSuccessListener(queryDocumentSnapshots -> {
+            ArrayList<FriendRequest> requests = new ArrayList<>();
+            for (DocumentSnapshot doc : queryDocumentSnapshots.getDocuments()) {
+                requests.add(doc.toObject(FriendRequest.class));
+            }
+            callback.onSuccess(requests);
+        }).addOnFailureListener(callback::onFailure);
+    }
+
+    /**
+     * add new friend request to the database
+     *
+     * @param fromUID
+     * @param toUID
+     * @param callback
+     */
+    public void sendFriendRequest(String fromUID, String toUID, final FirebaseCallback<Void> callback) {
+        FriendRequest friendRequest = new FriendRequest(new Date(), fromUID, false, toUID);
+        db.collection("requests").document().set(friendRequest)
+                .addOnSuccessListener(callback::onSuccess)
+                .addOnFailureListener(callback::onFailure);
+    }
+
+    /**
+     * flip the friend request (either approve or deny)
```

Comment:
I'm not sure how I feel about flip request. Having an approve and a deny method would be better

Reply:
It will have duplicated code for both of them, remember this is a callback-based library

Reply:
While I don't think we should change it for this project I would say the code duplication is worth it in this instance if this were a real project, since it makes the code more easy to understand and change in the future.